lib/zone.ts

```diff
@@ -1352,13 +1363,18 @@ const Zone: ZoneType = (function(global: any) {
           }
         }
       }
-      error.stack = error.zoneAwareStack = frames.join('\n');
+      if (this && this !== global) {
+        this.stack = this.zoneAwareStack = frames.join('\n');
+      } else {
+        error.stack = error.zoneAwareStack = frames.join('\n');
+        error.originalStack = originalStack;
+      }
     }
-    return error;
-  }
+    return (this && this !== global) ? this : error;
```

Comment:
The fact that sometimes we return `this` and sometimes `errror` concerns me. We have two seperate code path and I fear that things will only work in one or the other code path.